articles/ros2_security_contexts.md

```diff
@@ -0,0 +1,313 @@
+As such, additional tooling and extensions to SROS 2 are necessary to support this new paradigm.
+
+## Keystore
+
+With the additional structure of contexts, it’s perhaps best to take the opportunity to restructure the keystore layout as well.
+Rather than a flat directory of namespaced node security directories, we can push all such security directories down into a designated `contexts` sub-folder.
+Similarly, private and public keystore materials can also be pushed down into their own respective sub-folders within the root keystore directory.
+This is reminiscent of the pattern used earlier Keymint [1].
+
+
+```
+$ tree keystore/
+keystore
+├── contexts
+│   └── ...
+│       └── ...
+├── private
+│   ├── ca.csr.pem
+│   └── ca.key.pem
+└── public
+    ├── ca.cert.pem
+    ├── identity_ca.cert.pem
+    └── permissions_ca.cert.pem
+```
+
+
+### ``public``
+
+The ``public`` directory contains anything permissible as public, such as public certificates for the identity or permissions certificate authorities.
+As such, this can be given read access to all executables.
+Note that in the default case, both the `identity_ca` and `permissions_ca` are the same CA certificate.
+
+### ``private``
+
+The ``private`` directory contains anything permissable as private, such as private key material for aforementioned certificate authorities.
+This directory should be redacted before deploying the keystore onto the target device/robot.
+
+### ``contexts``
+
+The ``contexts`` directory contains the security artifacts associated with individual contexts, and thus node directories are no longer relevant.
+Similar to node directories however, the `contexts` folder may still recursively nest sub-paths for organizing separate contexts.
+The `ROS_SECURITY_ROOT_DIRECTORY` environment variable should by convention point to this directory.
+
+
+## Runtime
+
+Given the normative case where a context within a policy may be specific to a single node/container process, the namespace the node is remapped to will inevitably affect the required security permissions necessary within the context.
+To highlight this interdependency, and to help avoid context path collisions, a hierarchy borrowing namespaces is appropriate. 
+By convention, roslaunch could be used to prefix relative context paths for single process node or containers using the namespace in scope, to enable a convention of composable launch files with adjustable and parameterized context paths.
+Given the runtime command argument for specifying the fully qualified context path, roslaunch would accordly resolve relative context paths for executables, as defined by launch attributes.
+
+### Unqualified context path
+
+For single process nodes with unqualified context paths, the context directory will subsequently default to the root level context.
+
+``` xml
+<launch>
+  <node pkg="demo_nodes_cpp" exec="talker"/>
+  <node pkg="demo_nodes_cpp" exec="listener"/>
+</launch>
+```
```

Comment:
Considering the previous discussion, we can't do any assumption about how the context name/path relates to the node name. I would delete the whole discussion about launch files, and just put an example of how explicitly pass a context name:

``` xml
<launch>
  <node pkg="demo_nodes_cpp" exec="talker" context="talker"/>
  <node pkg="demo_nodes_cpp" exec="listener"  context="listener"/>
</launch>
```

If the context name/path isn't specified, we can take `"/"` or `""` as the default.

P.S.: There's a lot of improvement to do in ROS2 about launchfile reuse, e.g.: there's no recommended pattern to add components to a container, etc.

Reply:
Ah, I was still thinking, regardless of RCL only being able to take fully qualified context paths, ros launch could still be used to resolve relative paths to fully qualified paths before reaching RCL by using the namespace in scope (regardless of the namespace code may hardcode to), as just supporting convention of composing launch files. We can narrow this down at the security WG meeting today.